### ISSUE TITLE
How to use instance.evaluate in entry of State.

### ISSUE THREAD

**Issue by ogis-shinki** [steelbreeze/state#18]
Although I wrote instance.evaluate() in entry of State(see the below code), the instance can't move to a next state.  
Could you give me a way to evaluate in entry?

```javascript
var state = require("@steelbreeze/state");

// create event class that a transition will respond to
class MyEvent {
	constructor(fieldA, fieldB) { this.fieldA = fieldA; this.fieldB = fieldB; }

	toString() { return JSON.stringify(this); }
}

// log state entry, exit and trigger event evaluation
state.log.add(message => console.info(message), state.log.Entry | state.log.Exit | state.log.Evaluate);

// create the state machine model elements
const model = new state.State("model");
const initial = new state.PseudoState("initial", model, state.PseudoStateKind.Initial);
const stateA = new state.State("stateA", model);
const stateB = new state.State("stateB", model).entry(() => {console.log("hellkow B")}).exit(() => {console.log("bye B")});
let instance = new state.Instance("instance", model);

// HERE
stateA.entry(() => {
	instance.evaluate(new MyEvent(1, 2));
});

stateA.on(MyEvent).when((e) => { return e.fieldA === 1; }).to(stateB);
stateB.on(MyEvent).when((e) => { return e.fieldA === 1; }).to(stateA);
// create the transition from initial pseudo state to stateA
initial.to(stateA);


```

**Comment by mesmo** [comment]
Hi @ogis-shinki - the reason your call to instance.evaluate is not working is because the state machine instance is initialised in the ```Instance`` constructor; as the part of your model definition is declared after this, it is not available.

A couple of general suggestions:
1. As @stelbreeze/state is designed to allow multiple state machine instances that conform to a single state machine model, it is not advised to reference a specific instance in any of the entry/exit callbacks. If you need an instance object in these callbacks, I recommend you pass it as part of the trigger passed to Instance.evaluate (e.g. make it a member of your MyEvent class).
2. When a new instance is evaluated, its best to put minimal logic in place for your machine to get to a steady initial state and then use events from then on rather than doing so much in the initialisation.
3. @steelbreeze/state has its own transactions internally so calling the ```Instance.evaluate``` method inside the entry callback will have unpredictable behaviour as you're inside an uncommitted transaction when you call it. You are better off using continuation transitions to trigger the this sort of transition (perhaps putting a guard condition on the transition.

**Comment by ogis-shinki** [comment]
Thank you for your reply.  
As you said, my code is wrong. I resolved this problem by your reply.  

I'm concerned about one of your suggestions.

> @steelbreeze/state has its own transactions internally so calling the Instance.evaluate method inside the entry callback will have unpredictable behaviour as you're inside an uncommitted  

I think @steelbreeze/state should makes the Instance.evaluate method inside the entry callback deffered.

**Comment by mesmo** [comment]
Reworking the event queue (only currently used for deferred events) is something I can take a look into...

**Comment by redboltz** [comment]
Posting events during transition is typical issue in state machine library.
Here is C++ Boost.SML example.
https://github.com/boost-experimental/sml/issues/144

I think that event there are some ways to implement it.

1. Providing new API something like `post(event)`
2. (Re)use `evaluate()` and somehow check during transition or not in `evaluate()` and dispatch invoke immediately or en-queue.

I'd like to have this functionality.

**Comment by mesmo** [comment]
I will look to make it part of the evaluate call - to determine if there is an active transaction or not and on the basis of that, queue it or process it immediately.

**Comment by redboltz** [comment]
I'm not sure it is able to do but if we could have an optional parameter current `instance` in actions (entry/exit/transition effect), users implement posting event to the instance easily.

```js
stateA.entry((instance) => {
	instance.evaluate(new MyEvent(1, 2));
});
```

What do you think?

**Comment by mesmo** [comment]
Take a look at the latest code on the master branch; it now supports calls to ```Instance.evaluate``` in actions, for example:

```javascript
stateA.entry((trigger, instance) => {
	instance.evaluate(new MyEvent(1, 2));
});
```

The trigger event was always passed as the first parameter to the action callbacks; instance now added as a second parameter.

A call within an action adds the new trigger to the deferred event queue for for evaluation after the current run-to-completion step has completed.

**Comment by ogis-shinki** [comment]
Thank you.  
I will develop systems easier using @steelbreeze/state by the new feature.

**Comment by redboltz** [comment]
@mesmo , thank you. It's very useful :)

**Comment by mesmo** [comment]
Code published as v8.1.3
